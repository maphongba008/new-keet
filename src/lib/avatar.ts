import { ViewStyle } from 'react-native'
import b4a from 'b4a'

import { DEFAULT_AVATAR_COLORS } from './commonStyles'

export const SEED_BITS = 32
export const EMOJI_REGEXP = /\p{Emoji}/u

export const name2seed = (name: string) => {
  if (name.length === SEED_BITS) {
    return name
  } else if (name.length > SEED_BITS) {
    return name.substring(0, SEED_BITS - 1)
  }
  return name.padEnd(SEED_BITS)
}

function randInt(min: number, max: number, seed?: string): number {
  const rand = seed
    ? b4a.readUInt32LE(b4a.from(seed, 'hex')) / (2 ** SEED_BITS - 1)
    : Math.random()
  min = Math.ceil(min)
  max = Math.floor(max)
  return Math.floor(rand * (max - min)) + min
}

export const randomColor = (seed?: string) => {
  return DEFAULT_AVATAR_COLORS[randInt(0, DEFAULT_AVATAR_COLORS.length, seed)]
}

export function getScaledFontSize({
  width,
  height,
}: Pick<ViewStyle, 'width' | 'height'>): number {
  const baseWidth = 30
  const baseHeight = 30
  const baseFontSize = 12
  const maxFontThreshold = 40

  if (typeof height !== 'number' || typeof width !== 'number') {
    return baseFontSize
  }

  // Calculate scaling factor
  const scalingFactor = (width * height) / (baseWidth * baseHeight)

  // Scale font size
  const fontSize = baseFontSize * Math.sqrt(scalingFactor)
  const minThreshold = Math.max(fontSize, baseFontSize)

  return Math.min(minThreshold, maxFontThreshold)
}
