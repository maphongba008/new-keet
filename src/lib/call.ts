import { Alert } from 'react-native'

import { callLeave } from '@holepunchto/keet-store/store/call'

import type { Strings } from 'i18n/strings'

import { doVibrateSuccess } from './haptics'
import { getDispatch } from './store'

type showCallOnGoingAlertProps = {
  onPressLeave: Function
  onPressStay?: Function
  alertDesc?: string
  strings: Strings
}

export const showCallOnGoingAlert = ({
  onPressLeave,
  onPressStay,
  alertDesc,
  strings,
}: showCallOnGoingAlertProps) => {
  Alert.alert(
    strings.call.alertTitle,
    alertDesc || strings.call.alertDesc,
    [
      {
        text: strings.call.stayInCall,
        onPress: () => {
          onPressStay?.()
        },
        style: 'cancel',
      },
      {
        text: strings.call.leaveCall,
        onPress: () => {
          onPressLeave?.()
        },
        style: 'destructive',
      },
    ],
    { cancelable: true },
  )
}

export const dismissCall = () => {
  try {
    const dispatch = getDispatch()
    dispatch(callLeave())
    doVibrateSuccess()
  } catch (err) {
    console.log('Error in dismissCall', err)
  }
}
