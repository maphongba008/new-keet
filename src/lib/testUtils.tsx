import React, { PropsWithChildren } from 'react'
import { render, renderHook } from '@testing-library/react-native'
import type { RenderOptions } from '@testing-library/react-native'
import { Provider } from 'react-redux'
import createSagaMiddleware from 'redux-saga'
import { configureStore } from '@reduxjs/toolkit'

import { setBackend } from '@holepunchto/keet-store/backend'
import {
  APP_CONTEXT_KEY,
  ENV_CONTEXT_KEY,
  PLATFORM_CONTEXT_KEY,
} from '@holepunchto/keet-store/store/mobile'

import { injectStore } from './store'
import { createKeetStore, reducers } from '../store'

const mockSubscribe = {
  on: jest.fn(),
  data: jest.fn(),
  destroy: jest.fn(),
}
const MOCK_APP_CONTEXT_KEY = APP_CONTEXT_KEY
const MOCK_PLATFORM_CONTEXT_KEY = PLATFORM_CONTEXT_KEY
const MOCK_ENV_CONTEXT_KEY = ENV_CONTEXT_KEY

const mockedBackendApi = {
  api: {
    core: {
      getVersion: jest.fn(),
      subscribeRoomVersion: jest.fn().mockReturnValue(mockSubscribe),
      upgradeRoom: jest.fn(),
      getIdentity: jest.fn().mockReturnValue({
        local: {
          deviceId: 'device-id',
        },
        devices: [],
      }),
      subscribeIdentity: jest.fn().mockReturnValue(mockSubscribe),
      generateIdentity: jest.fn(),
      restoreIdentity: jest.fn(),
      confirmIdentity: jest.fn(),
      confirmPairingRequest: jest.fn(),
      subscribeIdentityPairingRequests: jest
        .fn()
        .mockReturnValue(mockSubscribe),
      updateIdentityProfile: jest.fn(),
      updateIdentityDevice: jest.fn(),
      startPairingIdentity: jest.fn(),
      pairIdentity: jest.fn(),
      createRoom: jest.fn(),
      pairRoom: jest.fn(),
      leaveRoom: jest.fn(),
      startPairingRoom: jest.fn(),
      stopPairingRoom: jest.fn(),
      subscribePendingRooms: jest.fn().mockReturnValue(mockSubscribe),
      subscribeActivateRoom: jest.fn().mockReturnValue(mockSubscribe),
      getRoomKeys: jest.fn().mockReturnValue({
        discoveryKey: 'Testkey',
      }),
      getTipSize: jest.fn(),
      bookmark: jest.fn(),
      unbookmark: jest.fn(),
      subscribeBookmarks: jest.fn().mockReturnValue(mockSubscribe),
      getRecentRooms: jest.fn().mockResolvedValue([]),
      getLocalRoomMetadata: jest.fn(),
      subscribeRecentRooms: jest.fn().mockReturnValue(mockSubscribe),
      subscribeUnreadActivity: jest.fn().mockReturnValue(mockSubscribe),
      setRoomUnread: jest.fn(),
      bumpRoomPriority: jest.fn(),
      addDevice: jest.fn(),
      removeDevice: jest.fn(),
      getDevice: jest.fn(),
      getDevices: jest.fn(),
      subscribeDevice: jest.fn().mockReturnValue(mockSubscribe),
      subscribeDevices: jest.fn().mockReturnValue(mockSubscribe),
      createIdentityInvitation: jest.fn(),
      createInvitation: jest.fn(),
      updateMember: jest.fn(),
      getMember: jest.fn().mockReturnValue({}),
      getMembers: jest.fn(),
      getModerators: jest.fn(),
      getMemberCount: jest.fn(),
      waitUntilJoined: jest.fn().mockReturnValue({
        memberId: 'testMember',
      }),
      getMemberForDevice: jest.fn(),
      subscribeMemberCount: jest.fn().mockReturnValue(mockSubscribe),
      subscribeMember: jest.fn().mockReturnValue(mockSubscribe),
      subscribeMembers: jest.fn().mockReturnValue(mockSubscribe),
      subscribeModerators: jest.fn().mockReturnValue(mockSubscribe),
      subscribeMemberForDevice: jest.fn().mockReturnValue(mockSubscribe),
      getSearchMembers: jest.fn(),
      getPushNotificationMembers: jest.fn(),
      subscribeSearchMembers: jest.fn().mockReturnValue(mockSubscribe),
      subscribePushNotificationMembers: jest
        .fn()
        .mockReturnValue(mockSubscribe),
      blockMember: jest.fn(),
      unblockMember: jest.fn(),
      getBlockedMembers: jest.fn(),
      subscribeBlockedMembers: jest.fn(),
      getConnectedMembers: jest.fn(),
      subscribeConnectedMembers: jest.fn().mockReturnValue(mockSubscribe),
      updateConfig: jest.fn(),
      getConfig: jest.fn(),
      subscribeConfig: jest.fn().mockReturnValue(mockSubscribe),
      addChatMessage: jest.fn(),
      updateChatMessage: jest.fn(),
      removeChatMessage: jest.fn(),
      addCallStartedEvent: jest.fn(),
      addRoomAvatarChangedEvent: jest.fn(),
      addEvent: jest.fn(),
      removeEvent: jest.fn(),
      getChatMessage: jest.fn(),
      getChatMessages: jest.fn(),
      getChatLength: jest.fn(),
      subscribeChatMessages: jest.fn().mockReturnValue(mockSubscribe),
      subscribeChatLength: jest.fn().mockReturnValue(mockSubscribe),
      sendInvitations: jest.fn(),
      updatePrivateMailboxRecord: jest.fn(),
      getPrivateMailboxRecords: jest.fn(),
      subscribePrivateMailbox: jest.fn().mockReturnValue(mockSubscribe),
      addReaction: jest.fn(),
      removeReaction: jest.fn(),
      getReactions: jest.fn(),
      subscribeReactions: jest.fn().mockReturnValue(mockSubscribe),
      addFileBlob: jest.fn(),
      addFile: jest.fn(),
      saveFileBlob: jest.fn(),
      saveFile: jest.fn(),
      sendFile: jest.fn(),
      removeFile: jest.fn(),
      subscribeFileInfo: jest.fn().mockReturnValue(mockSubscribe),
      clearAllFiles: jest.fn(),
      getFileEntry: jest.fn(),
      getLocalFiles: jest.fn(),
      startDownload: jest.fn(),
      cancelDownload: jest.fn(),
      clearFile: jest.fn(),
      unclearFile: jest.fn(),
      getFileCleared: jest.fn(),
      subscribeFileCleared: jest.fn().mockReturnValue(mockSubscribe),
      getStats: jest.fn().mockResolvedValue([]),
    },
    preferences: {
      query: jest.fn().mockResolvedValue({}),
      update: jest.fn(),
      subscribe: jest.fn().mockReturnValue({
        on: jest.fn(),
        data: jest.fn(),
      }),
    },
    mobile: {
      subscribeUnhandledErrors: jest.fn().mockReturnValue({
        on: jest.fn(),
        data: jest.fn(),
      }),
      getSwarmId: jest.fn().mockResolvedValue('swarmId'),
    },
  },
}
const mockedCallApi = {
  authorize: jest.fn(),
  deauthorize: jest.fn(),
  joinPeer: jest.fn(),
  leavePeer: jest.fn(),
  requestTrack: jest.fn(),
  unrequestTrack: jest.fn(),
  signalMedia: jest.fn(),
  getRequestedTracks: jest.fn(),
  subscribeRequestedTracks: jest.fn().mockReturnValue(mockSubscribe),
  subscribeMediaSignals: jest.fn().mockReturnValue(mockSubscribe),
  subscribeProxyUpdates: jest.fn().mockReturnValue(mockSubscribe),
  shareTrack: jest.fn(),
  unshareTrack: jest.fn(),
  relayTrack: jest.fn(),
  unrelayTrack: jest.fn(),
  subscribeAvailableTracks: jest.fn().mockReturnValue(mockSubscribe),
  getAvailableTracks: jest.fn(),
  setPresent: jest.fn(),
  setNotPresent: jest.fn(),
  presence: {
    subscribePresence: jest.fn().mockReturnValue(mockSubscribe),
    getPresence: jest.fn(),
  },
  devices: {
    subscribeDevices: jest.fn().mockReturnValue(mockSubscribe),
  },
}

jest.mock('redux-saga/effects', () => {
  return {
    ...jest.requireActual('redux-saga/effects'),
    getContext: jest.fn().mockImplementation((key: any) => {
      const map = {
        [MOCK_ENV_CONTEXT_KEY]: { isTest: true, isMac: true, isMobile: true },
        [MOCK_APP_CONTEXT_KEY]: {
          localeManager: {
            setLocale: jest.fn(),
          },
        },
        [MOCK_PLATFORM_CONTEXT_KEY]: {
          config: {
            getInitialURL: jest.fn().mockResolvedValue(''),
          },
          media: {
            status: {
              microphone: jest.fn().mockResolvedValue('granted'),
              camera: jest.fn().mockResolvedValue('granted'),
              screen: jest.fn().mockResolvedValue('unsupported'),
            },
            access: {
              microphone: jest.fn().mockResolvedValue(true),
              camera: jest.fn().mockResolvedValue(true),
              screen: jest.fn().mockResolvedValue(false),
            },
          },
        },
      }
      return map[key]
    }),
  }
})

export const mockCoreApis = () =>
  setBackend({ backendApi: mockedBackendApi, callApi: mockedCallApi })

export const getMockStore = (preloadedState: Object = {}) => {
  mockCoreApis()
  const store = createKeetStore(preloadedState)
  return store
}

export const getMockStoreWithMiddleware = () => {
  const sagaMiddleware = createSagaMiddleware()
  const store = configureStore({
    reducer: reducers,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: false,
      }).concat([sagaMiddleware]),
  })
  injectStore(store)
  return { store, sagaMiddleware }
}

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: any
  store?: any
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = getMockStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export function renderHookWithProviders(
  hook: () => unknown,
  {
    preloadedState = {},
    store = getMockStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
