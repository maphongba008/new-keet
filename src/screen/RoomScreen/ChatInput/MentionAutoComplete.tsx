import { memo, useCallback, useMemo } from 'react'
import { ListRenderItem, Pressable, StyleSheet, Text, View } from 'react-native'
import { shallowEqual, useSelector } from 'react-redux'
import isEqual from 'react-fast-compare'
import { FlatList } from 'react-native-gesture-handler'

import { getAppCurrentRoomId } from '@holepunchto/keet-store/store/app'
import {
  getMemberBlockedById,
  getMembersBlockedByRoomId,
} from '@holepunchto/keet-store/store/member'

import { getChatBarHeight } from 'reducers/application'

import { MemberAvatar } from 'component/Avatar'
import MemberTag from 'component/MemberTag'
import { createThemedStylesheet } from 'component/theme'
import s, { UI_SIZE_4, UI_SIZE_8 } from 'lib/commonStyles'
import { normalize } from 'lib/hooks/useMember'
import { getRoomTypeFlags, useRoom } from 'lib/hooks/useRoom'
import { SearchProfile } from 'lib/types'

const MENTION_ITEM_HEIGHT = 30
const MentionItem = ({
  item,
  onSelectMentionProfile,
  roomId,
}: {
  item: SearchProfile
  onSelectMentionProfile: (name: string) => void
  roomId: string
}) => {
  const styles = getStyles()
  const blocked = useSelector(
    (state) => getMemberBlockedById(state, roomId, item.memberId),
    shallowEqual,
  )
  const { isDm } = getRoomTypeFlags(useRoom(roomId)?.roomType)
  const member = normalize({ member: item, blocked })
  const textStyles: any = isDm ? {} : member?.theme
  const selectMentionProfile = useCallback(() => {
    onSelectMentionProfile(item.displayName)
  }, [item.displayName, onSelectMentionProfile])

  if (member.blocked) {
    return null
  }

  return (
    <Pressable
      testID={item.memberId}
      accessibilityLabel={item.memberId}
      style={[styles.autocompleteListItem, s.centerAlignedRow]}
      onPress={selectMentionProfile}
      pointerEvents="box-only"
    >
      <MemberAvatar member={member} />
      <Text style={[styles.autocompleteListItemText, textStyles]}>
        {item.displayName}
      </Text>
      {!isDm && (
        <MemberTag member={member} isList containerStyleProps={styles.tag} />
      )}
    </Pressable>
  )
}

const MentionItemSeparator = () => {
  const styles = getStyles()

  return <View style={styles.autocompleteSeparator} />
}

const MentionsAutocomplete = ({
  profiles,
  onSelectMentionProfile,
}: {
  profiles: SearchProfile[]
  onSelectMentionProfile: (name: string) => void
}) => {
  const roomId: string = useSelector(getAppCurrentRoomId)
  const styles = getStyles()
  const blockedMembers = useSelector(getMembersBlockedByRoomId(roomId))
  const filteredProfiles = useMemo(
    () => profiles.filter((profile) => !blockedMembers[profile.memberId]),
    [blockedMembers, profiles],
  )
  const chatBarHeight = useSelector(getChatBarHeight)

  const renderItem: ListRenderItem<SearchProfile> = useCallback(
    ({ item }) => {
      return <MentionItem {...{ roomId, item, onSelectMentionProfile }} />
    },
    [onSelectMentionProfile, roomId],
  )
  const getItemLayout = useCallback(
    (_: unknown, index: number) => ({
      length: MENTION_ITEM_HEIGHT,
      offset: MENTION_ITEM_HEIGHT * index,
      index,
    }),
    [],
  )

  if (!filteredProfiles.length) return null

  return (
    <View
      style={[styles.autocompleteRoot, { bottom: chatBarHeight + UI_SIZE_4 }]}
    >
      <FlatList
        data={filteredProfiles}
        contentContainerStyle={styles.flatlist}
        renderItem={renderItem}
        ItemSeparatorComponent={MentionItemSeparator}
        getItemLayout={getItemLayout}
        keyboardShouldPersistTaps="always"
        nestedScrollEnabled
      />
    </View>
  )
}

const getStyles = createThemedStylesheet((theme) => {
  const padding = theme.spacing.standard / 2
  const styles = StyleSheet.create({
    autocompleteListItem: {
      height: MENTION_ITEM_HEIGHT,
    },
    autocompleteListItemText: {
      ...theme.text.body,
      paddingLeft: padding,
    },
    autocompleteRoot: {
      backgroundColor: theme.color.grey_700,
      borderColor: theme.color.bg4,
      borderRadius: theme.border.radiusNormal * 2,
      borderWidth: 1,
      left: padding,
      maxHeight: 200,
      overflow: 'hidden',
      position: 'absolute',
      width: '100%',
    },
    autocompleteSeparator: {
      marginTop: UI_SIZE_4,
    },
    flatlist: {
      padding: UI_SIZE_8,
    },
    tag: {
      marginLeft: UI_SIZE_4,
    },
  })
  return styles
})

export default memo(MentionsAutocomplete, isEqual)
