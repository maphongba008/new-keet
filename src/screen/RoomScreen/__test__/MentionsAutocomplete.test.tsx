import React from 'react'
import { fireEvent, screen } from '@testing-library/react-native'

import { renderWithProviders } from 'lib/testUtils'

import MentionAutoComplete from '../ChatInput/MentionAutoComplete'

describe('test MentionsAutocomplete', () => {
  it('render correctly', async () => {
    const onSelectProfile = jest.fn()
    const element = (
      <MentionAutoComplete
        onSelectMentionProfile={onSelectProfile}
        profiles={[
          {
            memberId: '1',
            name: 'member 1',
            displayName: 'Member 1',
            avatarUrl: null,
          },
          {
            memberId: '2',
            name: 'member 2',
            displayName: 'Member 2',
            avatarUrl: 'base64',
            capabilities: 31, // Admin
          },
        ]}
      />
    )
    renderWithProviders(element, {
      preloadedState: {
        app: {
          currentRoomId: 'room1',
        },
        room: {
          roomById: {
            room1: {
              roomType: '1',
            },
          },
        },
        member: {
          blockedMembersByRoomId: {
            room1: { '1': true },
          },
        },
      },
    })
    expect(screen.toJSON()).toMatchSnapshot()
    // expect member 1 was not rendered due to blocked
    expect(screen.queryByTestId('1')).toBeFalsy()
    // click on member 2
    const item = await screen.findByTestId('2')
    fireEvent.press(item)
    expect(onSelectProfile).toHaveBeenCalledWith('Member 2')
  })

  it('For Dm room render without member role color and tag', async () => {
    const onSelectProfile = jest.fn()
    const element = (
      <MentionAutoComplete
        onSelectMentionProfile={onSelectProfile}
        profiles={[
          {
            memberId: '1',
            name: 'member 1',
            displayName: 'Member 1',
            avatarUrl: null,
          },
          {
            memberId: '2',
            name: 'member 2',
            displayName: 'Member 2',
            avatarUrl: 'base64',
          },
        ]}
      />
    )
    renderWithProviders(element, {
      preloadedState: {
        app: {
          currentRoomId: 'room1',
        },
        room: {
          roomById: {
            room1: {
              roomType: '2',
            },
          },
        },
        member: {
          blockedMembersByRoomId: {
            room1: [],
          },
        },
      },
    })
    expect(screen.toJSON()).toMatchSnapshot()
  })

  it('render null with no profiles', async () => {
    const onSelectProfile = jest.fn()
    const element = (
      <MentionAutoComplete
        onSelectMentionProfile={onSelectProfile}
        profiles={[]}
      />
    )
    renderWithProviders(element, {
      preloadedState: {
        app: {
          currentRoomId: 'room1',
        },
        room: {
          roomById: {
            room1: {
              roomType: '1',
            },
          },
        },
        member: {
          blockedMembersByRoomId: {
            room1: [],
          },
        },
      },
    })
    expect(screen.toJSON()).toMatchSnapshot()
  })
})
