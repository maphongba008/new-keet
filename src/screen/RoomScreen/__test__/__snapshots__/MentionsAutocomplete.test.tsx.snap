// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MentionsAutocomplete For Dm room render without member role color and tag 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#1A1D29",
        "borderColor": "#323A4C",
        "borderRadius": 16,
        "borderWidth": 1,
        "left": 8,
        "maxHeight": 200,
        "overflow": "hidden",
        "position": "absolute",
        "width": "100%",
      },
      {
        "bottom": 4,
      },
    ]
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    collapsable={false}
    contentContainerStyle={
      {
        "padding": 8,
      }
    }
    data={
      [
        {
          "avatarUrl": null,
          "displayName": "Member 1",
          "memberId": "1",
          "name": "member 1",
        },
        {
          "avatarUrl": "base64",
          "displayName": "Member 2",
          "memberId": "2",
          "name": "member 2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    handlerTag={2}
    handlerType="NativeViewGestureHandler"
    keyExtractor={[Function]}
    keyboardShouldPersistTaps="always"
    nestedScrollEnabled={true}
    onContentSizeChange={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
    waitFor={
      [
        {
          "current": null,
        },
        {
          "current": null,
        },
      ]
    }
  >
    <View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        <View
          accessibilityLabel="1"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          pointerEvents="box-only"
          style={
            [
              {
                "height": 30,
              },
              {
                "alignItems": "center",
                "flexDirection": "row",
              },
            ]
          }
          testID="1"
        >
          <View
            accessibilityLabel="avatar_image"
            accessible={true}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "hsla(49,70%,70%,0.9)",
                "borderRadius": 15,
                "height": 30,
                "justifyContent": "center",
                "width": 30,
              }
            }
            testID="avatar_image"
          >
            <Text
              style={
                [
                  {
                    "color": "#ffffff",
                    "fontFamily": "OpenSans-Bold",
                    "fontSize": 12,
                    "includeFontPadding": false,
                  },
                  {
                    "fontSize": 12,
                  },
                ]
              }
            >
              M
            </Text>
          </View>
          <Text
            style={
              [
                {
                  "color": "#ffffff",
                  "fontFamily": "OpenSans-Regular",
                  "fontSize": 16,
                  "includeFontPadding": false,
                  "paddingLeft": 8,
                },
                {},
              ]
            }
          >
            Member 1
          </Text>
        </View>
        <View
          style={
            {
              "marginTop": 4,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        <View
          accessibilityLabel="2"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          pointerEvents="box-only"
          style={
            [
              {
                "height": 30,
              },
              {
                "alignItems": "center",
                "flexDirection": "row",
              },
            ]
          }
          testID="2"
        >
          <Image
            accessibilityLabel="avatar_image"
            accessible={true}
            cachePolicy="memory"
            contentFit="contain"
            recyclingKey="base64"
            source={
              {
                "uri": "base64",
              }
            }
            style={
              {
                "borderRadius": 15,
                "height": 30,
                "width": 30,
              }
            }
            testID="avatar_image"
          />
          <Text
            style={
              [
                {
                  "color": "#ffffff",
                  "fontFamily": "OpenSans-Regular",
                  "fontSize": 16,
                  "includeFontPadding": false,
                  "paddingLeft": 8,
                },
                {},
              ]
            }
          >
            Member 2
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`test MentionsAutocomplete render correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#1A1D29",
        "borderColor": "#323A4C",
        "borderRadius": 16,
        "borderWidth": 1,
        "left": 8,
        "maxHeight": 200,
        "overflow": "hidden",
        "position": "absolute",
        "width": "100%",
      },
      {
        "bottom": 4,
      },
    ]
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    collapsable={false}
    contentContainerStyle={
      {
        "padding": 8,
      }
    }
    data={
      [
        {
          "avatarUrl": "base64",
          "capabilities": 31,
          "displayName": "Member 2",
          "memberId": "2",
          "name": "member 2",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    keyExtractor={[Function]}
    keyboardShouldPersistTaps="always"
    nestedScrollEnabled={true}
    onContentSizeChange={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
    waitFor={
      [
        {
          "current": null,
        },
        {
          "current": null,
        },
      ]
    }
  >
    <View>
      <View
        onFocusCapture={[Function]}
        style={null}
      >
        <View
          accessibilityLabel="2"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          pointerEvents="box-only"
          style={
            [
              {
                "height": 30,
              },
              {
                "alignItems": "center",
                "flexDirection": "row",
              },
            ]
          }
          testID="2"
        >
          <Image
            accessibilityLabel="avatar_image"
            accessible={true}
            cachePolicy="memory"
            contentFit="contain"
            recyclingKey="base64"
            source={
              {
                "uri": "base64",
              }
            }
            style={
              {
                "borderRadius": 15,
                "height": 30,
                "width": 30,
              }
            }
            testID="avatar_image"
          />
          <Text
            style={
              [
                {
                  "color": "#ffffff",
                  "fontFamily": "OpenSans-Regular",
                  "fontSize": 16,
                  "includeFontPadding": false,
                  "paddingLeft": 8,
                },
                {
                  "color": "#53EA98",
                },
              ]
            }
          >
            Member 2
          </Text>
          <View
            style={
              [
                {
                  "marginLeft": 4,
                },
                [
                  {
                    "alignItems": "center",
                    "borderRadius": 6,
                    "justifyContent": "center",
                    "minHeight": 16,
                    "minWidth": 32,
                  },
                  {
                    "backgroundColor": "rgba(83, 234, 152, 0.2)",
                  },
                ],
              ]
            }
          >
            <Text
              allowFontScaling={false}
              style={
                [
                  {
                    "color": "#ffffff",
                    "fontFamily": "OpenSans-SemiBold",
                    "fontSize": 12,
                    "includeFontPadding": false,
                    "lineHeight": 17,
                    "paddingHorizontal": 4,
                    "textAlignVertical": "center",
                  },
                  {
                    "color": "#53EA98",
                  },
                  {
                    "fontSize": 10,
                    "lineHeight": 15,
                  },
                ]
              }
            >
              Admin
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`test MentionsAutocomplete render null with no profiles 1`] = `null`;
